File: SkyScanner/SkyScanner/FSCurveButton.m
//
//  FSCurveButton.m
//  SkyScanner
//
//  Created by 翁志方 on 16/8/11.
//  Copyright © 2016年 翁志方. All rights reserved.
//

#import "FSCurveButton.h"

@implementation FSCurveButton

- (instancetype)initWithFrame:(CGRect)frame curveSize:(FSCurveButtonSize)curveSize
{
    self = [super initWithFrame:frame];
    
    self.backgroundColor = [UIColor clearColor];
    
    CAShapeLayer *layer = [CAShapeLayer layer];
    CGMutablePathRef path = CGPathCreateMutable();
    
    CGFloat width = frame.size.width;
    CGFloat height = frame.size.height;
    
    CGFloat radius = width>height ? height : width;
    radius/=4.5;
    
    if (self.curveButtonSize == kFSCurveButtonSizeLeft) {

        CGPathMoveToPoint(path, nil, 0, height);
        
        CGPathAddArcToPoint(path, nil, 0, 0, width, 0, radius);
        
        CGPathAddLineToPoint(path, nil, width, 0);
        
        CGPathAddArcToPoint(path, nil, width, height, 0, height, radius);
        
        CGPathAddLineToPoint(path, nil, 0, height);
        
    }else{
        
        CGPathMoveToPoint(path, nil, 0, 0);
        
        CGPathAddArcToPoint(path, nil, width, 0, width, height, radius);
        
        CGPathAddLineToPoint(path, nil, 0, height);
        
        CGPathAddArcToPoint(path, nil, 0, 0, width, 0, radius);
        
        CGPathAddLineToPoint(path, nil, 0, height);
    }

    layer.path = path;
    layer.fillColor = [UIColor colorWithWhite:1 alpha:0.8].CGColor;
    
    [self.layer addSublayer:layer];

    return self;
}



@end

File: SkyScanner swift/SkyScanner/FSCurveButton.swift
//
//  FSCurveButton.swift
//  SkyScanner
//
//  Created by 翁志方 on 16/8/12.
//  Copyright © 2016年 翁志方. All rights reserved.
//

import UIKit

enum FSCurveButtonSizeType {
    case left,right
}


class FSCurveButton: UIButton {

    var curveButtonSize = FSCurveButtonSizeType.left;
    var fillColor : UIColor {
        get{
            return UIColor.init(cgColor: shapeLayer.fillColor!)
        }
        set{
            shapeLayer.fillColor = newValue.cgColor
        
        }
    }
    var shapeLayer = CAShapeLayer()
    
    init(frame:CGRect,curveType:FSCurveButtonSizeType) {
        super.init(frame: frame)
        
        let path = CGMutablePath()
        
        let width = frame.size.width
        let height = frame.size.height
        
        var radius = width>height ? height : width
        radius *= 0.3;
        
        if curveType == FSCurveButtonSizeType.left {
        
            path.move(to: CGPoint.init(x: 0, y: height))
            
            path.addArc(tangent1End: CGPoint.init(x: 0, y: 0), tangent2End:  CGPoint.init(x: width, y: 0), radius: radius)
            
            path.addLine(to: CGPoint.init(x: width, y: 0))
            
            path.addArc(tangent1End: CGPoint.init(x: width, y: height), tangent2End:  CGPoint.init(x: 0, y: height), radius: radius)
            
            path.addLine(to: CGPoint.init(x: 0, y: height))
            
        }else{
            
            path.move(to: CGPoint.init(x: 0, y: 0))
            
            path.addArc(tangent1End: CGPoint.init(x: width, y: 0), tangent2End:  CGPoint.init(x: width, y: height), radius: radius)
            
            
            path.addLine(to: CGPoint.init(x: width, y: height))
            
            path.addArc(tangent1End: CGPoint.init(x: 0, y: height), tangent2End:  CGPoint.init(x: 0, y: 0), radius: radius)
            
            path.addLine(to: CGPoint.init(x: 0, y: 0))

        }
        
        shapeLayer.path = path;
        
        shapeLayer.fillColor = UIColor.red.cgColor;
        
        self.layer.addSublayer(shapeLayer);
        
        
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    
    
    

}
